### PR TITLE
 Add pip resolver

### LINKED ISSUE #1642: Workflow linter to analyse PyPI library
Would share significant codebase with:
- https://github.com/databrickslabs/ucx/issues/1640

### PR DESCRIPTION
**Pull request by JCZuurmond**
## Changes
Add a pip resolver for later use by the `ImportResolver`. The resolver makes a library available to the path look up by i) installing the library and ii) augmenting the path lookup.

### Linked issues
Part of #1642
Follow up on #1694 

### Functionality 

- [ ] added relevant user documentation
- [ ] added new CLI command
- [ ] modified existing command: `databricks labs ucx ...`
- [ ] added a new workflow
- [x] modified existing workflow: `...`
- [ ] added a new table
- [ ] modified existing table: `...`

### Tests

- [ ] manually tested
- [x] added unit tests
- [ ] added integration tests
- [ ] verified on staging environment (screenshot attached)

### CODE REVIEW

**Review by nfx — DISMISSED** [review]
Why do we need to introduce new concepts and increase levels of indirection? Please provide trade-off analysis matrix with pros and cons of each approach.

**Comment by JCZuurmond** [comment]
> Why do we need to introduce new concepts and increase levels of indirection? Please provide trade-off analysis matrix with pros and cons of each approach.

Added motivation to the first PR comment

File: src/databricks/labs/ucx/contexts/application.py [hunk 1]
```diff
@@ -391,7 +395,7 @@ def file_resolver(self):
 
     @cached_property
     def dependency_resolver(self):
-        import_resolvers = [self.file_resolver, self.whitelist_resolver]
+        import_resolvers = [self.pip_resolver, self.file_resolver, self.whitelist_resolver]
```
Comment (ericvergnaud) [inline src/databricks/labs/ucx/contexts/application.py:398]: The pip resolver is NOT an import resolver. You need to add a _library_resolver member in DependencyResolver with type BaseLibraryResolver

Comment (ericvergnaud) [reply src/databricks/labs/ucx/contexts/application.py:398]: And PipResolver should derive from BaseLibraryResolver

File: src/databricks/labs/ucx/source_code/graph.py [hunk 2]
```diff
@@ -276,6 +276,11 @@ def resolve_import(self, path_lookup: PathLookup, name: str) -> MaybeDependency:
         assert self._next_resolver is not None
         return self._next_resolver.resolve_import(path_lookup, name)
 
+    def resolve_library_pip(self, path_lookup: PathLookup, name: str) -> MaybeDependency:
```
Comment (ericvergnaud) [inline src/databricks/labs/ucx/source_code/graph.py:279]: No, this goes against the rationale provided in the PR: introduces confusion and complexity

File: src/databricks/labs/ucx/source_code/graph.py [hunk 3]
```diff
@@ -295,6 +300,9 @@ def with_next_resolver(self, resolver: BaseImportResolver) -> BaseImportResolver
     def resolve_import(self, path_lookup: PathLookup, name: str) -> MaybeDependency:
         return self._fail('import-not-found', f"Could not locate import: {name}")
 
+    def resolve_library_pip(self, path_lookup: PathLookup, name: str) -> MaybeDependency:
```
Comment (ericvergnaud) [inline src/databricks/labs/ucx/source_code/graph.py:303]: no, see above

File: src/databricks/labs/ucx/source_code/site_packages.py [hunk 4]
```diff
@@ -1,7 +1,50 @@
...
+import tempfile
+from functools import cached_property
 from pathlib import Path
+from subprocess import CalledProcessError
+
+from databricks.labs.ucx.source_code.graph import BaseImportResolver, DependencyProblem, MaybeDependency
+from databricks.labs.ucx.source_code.path_lookup import PathLookup
+
+
+class PipResolver(BaseImportResolver):
```
Comment (ericvergnaud) [inline src/databricks/labs/ucx/source_code/site_packages.py:14]: no, you need to derive from BaseLibraryResolver

File: tests/unit/source_code/test_graph.py [hunk 5]
```diff
@@ -0,0 +1,26 @@
+from pathlib import Path
+
+from databricks.labs.ucx.source_code.files import FileLoader
+from databricks.labs.ucx.source_code.graph import Dependency, DependencyGraph, DependencyResolver, StubImportResolver
+from databricks.labs.ucx.source_code.notebooks.loaders import NotebookResolver, NotebookLoader
+from databricks.labs.ucx.source_code.site_packages import PipResolver
+
+
+def test_dependency_graph_register_library(mock_path_lookup):
```
Comment (ericvergnaud) [inline tests/unit/source_code/test_graph.py:9]: `registers` not `register`

File: tests/unit/source_code/test_jobs.py [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+    dependency = Dependency(FileLoader(), Path("test"))
+    notebook_loader = NotebookLoader()
+    notebook_resolver = NotebookResolver(notebook_loader)
+    import_resolver = PipResolver()
+    dependency_resolver = DependencyResolver(notebook_resolver, [import_resolver], mock_path_lookup)
+    dependency_graph = DependencyGraph(dependency, None, dependency_resolver, mock_path_lookup)
+    return dependency_graph
+
+
+def test_workflow_task_container_build_dependency_graph_not_yet_implemented(mock_path_lookup, graph):
```
Comment (ericvergnaud) [inline tests/unit/source_code/test_jobs.py:26]: "...reports_not_yet_implemented_problem"

File: tests/unit/source_code/test_jobs.py [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+
+    workflow_task_container = WorkflowTaskContainer(ws, task)
+    problems = workflow_task_container.build_dependency_graph(graph)
+
+    assert len(problems) == 5
+    assert all(problem.code == "not-yet-implemented" for problem in problems)
+    ws.assert_not_called()
+
+
+def test_workflow_task_container_build_dependency_graph_empty_task(mock_path_lookup, graph):
```
Comment (ericvergnaud) [inline tests/unit/source_code/test_jobs.py:41]: `builds` not `build`

File: tests/unit/source_code/test_jobs.py [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+    """No dependency problems with empty task"""
```
Comment (ericvergnaud) [inline tests/unit/source_code/test_jobs.py:42]: remove

File: tests/unit/source_code/test_jobs.py [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+    notebook_loader = NotebookLoader()
+    notebook_resolver = NotebookResolver(notebook_loader)
+    import_resolver = PipResolver()
+    dependency_resolver = DependencyResolver(notebook_resolver, [import_resolver], mock_path_lookup)
+    dependency_graph = DependencyGraph(dependency, None, dependency_resolver, mock_path_lookup)
+    return dependency_graph
+
+
+def test_workflow_task_container_build_dependency_graph_not_yet_implemented(mock_path_lookup, graph):
+    """Receive not yet implemented problems"""
```
Comment (ericvergnaud) [inline tests/unit/source_code/test_jobs.py:27]: remove

File: tests/unit/source_code/test_jobs.py [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+    task = jobs.Task(task_key="test")
+
+    workflow_task_container = WorkflowTaskContainer(ws, task)
+    problems = workflow_task_container.build_dependency_graph(graph)
+
+    assert len(problems) == 0
+    ws.assert_not_called()
+
+
+def test_workflow_task_container_build_dependency_graph_pytest_pypi_library(mock_path_lookup, graph):
```
Comment (ericvergnaud) [inline tests/unit/source_code/test_jobs.py:53]: ` builds`  not `build`

File: tests/unit/source_code/test_jobs.py [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+    """Install pytest PyPi library"""
```
Comment (ericvergnaud) [inline tests/unit/source_code/test_jobs.py:54]: remove

File: tests/unit/source_code/test_jobs.py [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+    workflow_task_container = WorkflowTaskContainer(ws, task)
+    problems = workflow_task_container.build_dependency_graph(graph)
+
+    assert len(problems) == 0
+    assert graph.path_lookup.resolve(Path("pytest")).exists()
+    ws.assert_not_called()
+
+
+def test_workflow_task_container_build_dependency_graph_unknown_pypi_library(mock_path_lookup, graph):
+    """Install unknown PyPi library"""
```
Comment (ericvergnaud) [inline tests/unit/source_code/test_jobs.py:68]: remove

File: tests/unit/source_code/test_jobs.py [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+
+    workflow_task_container = WorkflowTaskContainer(ws, task)
+    problems = workflow_task_container.build_dependency_graph(graph)
+
+    assert len(problems) == 0
+    assert graph.path_lookup.resolve(Path("pytest")).exists()
+    ws.assert_not_called()
+
+
+def test_workflow_task_container_build_dependency_graph_unknown_pypi_library(mock_path_lookup, graph):
```
Comment (ericvergnaud) [inline tests/unit/source_code/test_jobs.py:67]: `builds` not `build`

File: tests/unit/source_code/test_site_packages.py [hunk 14]
```diff
@@ -1,7 +1,27 @@
-from databricks.labs.ucx.source_code.site_packages import SitePackages
+from pathlib import Path
+
+from databricks.labs.ucx.source_code.site_packages import PipResolver, SitePackages
 from tests.unit import locate_site_packages
 
 
+def test_pip_resolver_library(mock_path_lookup):
```
Comment (ericvergnaud) [inline tests/unit/source_code/test_site_packages.py:7]: rename to test_pip_resolver_resolves_library

File: tests/unit/source_code/test_site_packages.py [hunk 15]
```diff
@@ -1,7 +1,27 @@
...
+    pip_installer = PipResolver()
+    maybe = pip_installer.resolve_library_pip(mock_path_lookup, "pytest")
+
+    assert len(maybe.problems) == 0
+    assert mock_path_lookup.resolve(Path("pytest")).exists()
+
+
+def test_pip_resolver_unknown_library(mock_path_lookup):
```
Comment (ericvergnaud) [inline tests/unit/source_code/test_site_packages.py:15]: test_pip_resolver_does_not_resolve_unknown_library

**Review by ericvergnaud — CHANGES_REQUESTED** [review]

File: src/databricks/labs/ucx/source_code/graph.py [hunk 16]
```diff
@@ -45,12 +45,12 @@ def dependency(self):
...
         return self._dependency.path
 
-    def register_library(self, name: str) -> MaybeGraph:
-        # TODO: use DistInfoResolver to load wheel/egg/pypi dependencies
-        # TODO: https://github.com/databrickslabs/ucx/issues/1642
-        # TODO: https://github.com/databrickslabs/ucx/issues/1643
-        # TODO: https://github.com/databrickslabs/ucx/issues/1640
-        return MaybeGraph(None, [DependencyProblem('not-yet-implemented', f'Library dependency: {name}')])
+    def register_library(self, library: str) -> list[DependencyProblem]:
+        maybe = self._resolver.resolve_library(self.path_lookup, library)
```
Comment (ericvergnaud) [inline src/databricks/labs/ucx/source_code/graph.py:49]: Don't remove the TODOs, there's still work to be done here, just not as part of this PR

**Review by ericvergnaud — CHANGES_REQUESTED** [review]

File: tests/unit/source_code/test_jobs.py (hunk 13 above)
Comment (ericvergnaud) [reply tests/unit/source_code/test_jobs.py:67]: is it not a duplicate of test_pip_resolver_unknown_library ?

**Review by ericvergnaud — APPROVED** [review]

**Pull request merged**